### PR TITLE
Sparse Adam optimizer for sparse gradients

### PR DESCRIPTION
**Pull request by ssnl**
1. Implements Lazy Adam optimizer. See details in #1285 .

2. Adds various checks to other optimizers that do not support sparse gradients.

3. Adds test for sparse gradients with SGD

cc: @ezyang

### CODE REVIEW

**Review by apaszke — COMMENTED** [review]
Should we e.g. keep the number of steps since the last update for each of the weight row separately, and once it gets the gradient, we would decay all the moments and apply the weight decay with stronger coefficients (i.e. decay with `decay^num_inactive_steps`). Not sure if this is what other frameworks do, or what people care about, but it seems to me as if that would more closely mimic the regular Adam optimizer.

**Comment by SsnL** [comment]
@apaszke Currently, Adagrad, which has custom code for sparse grad, disables weight decay for sparse grad. I think weight decay should be a at least method in base Optimizer as it is common in most (all?) optimizers. I am willing to do that in this PR (or in a new PR).

File: torch/optim/lazy_adam.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+                    return constructor(grad_indices, values, size)
+
+                exp_avg, exp_avg_sq = state['exp_avg'], state['exp_avg_sq']
+                beta1, beta2 = group['betas']
+
+                # Decay the first and second moment running average coefficient
+                #      old <- b * old + (1 - b) * new
+                # <==> old += (1 - b) * (new - old)
+                old_exp_avg_values = exp_avg._sparse_mask(grad)._values()
+                exp_avg.add_(make_sparse(grad_values.sub_(old_exp_avg_values).mul_(1 - beta1)))
```
Comment (apaszke) [inline torch/optim/lazy_adam.py]: this looks quite unsafe. You're modifying `grad_values` in-place. Doesn't this change the grad as well?

File: torch/optim/lazy_adam.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                old_exp_avg_sq_values = exp_avg_sq._sparse_mask(grad)._values()
+                exp_avg_sq.add_(make_sparse(grad_values_sq.sub_(old_exp_avg_sq_values).mul_(1 - beta2)))
```
Comment (apaszke) [inline torch/optim/lazy_adam.py]: same as above

File: torch/optim/lazy_adam.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+                numer = exp_avg._sparse_mask(grad)._values()
```
Comment (apaszke) [inline torch/optim/lazy_adam.py]: Are `self` values used in `._sparse_mask(tensor)` call, or is it just using the indices? If so, you're doing this gather on `grad` multiple times, while you could have done it only once

File: torch/optim/lazy_adam.py (hunk 2 above)
Comment (SsnL) [reply torch/optim/lazy_adam.py]: This should be okay, as the `grad_values_sq` is a new tensor.

File: torch/optim/lazy_adam.py (hunk 1 above)
Comment (SsnL) [reply torch/optim/lazy_adam.py]: Good point.

File: torch/optim/lazy_adam.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                    # Exponential moving average of squared gradient values
+                    state['exp_avg_sq'] = p.data.new().resize_as_(p.data).zero_()
+
+                state['step'] += 1
+
+                grad = grad.coalesce()  # the update is non-linear so indices must be unique
+                grad_indices = grad._indices()
+                grad_values = grad._values()
+                grad_values_sq = grad_values.pow(2)
+                size = torch.Size([x for x in grad.size()])
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/optim/lazy_adam.py]: This is following the sparse adagrad in https://github.com/pytorch/pytorch/blob/master/torch/optim/adagrad.py#L71. I'm not sure why `grad.size()` on't work, should I change both locations?

File: torch/optim/lazy_adam.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+                def make_sparse(values):
+                    constructor = type(p.grad.data)
+                    if grad_indices.dim() == 0 or values.dim() == 0:
+                        return constructor()
+                    return constructor(grad_indices, values, size)
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/optim/lazy_adam.py]: This is following the sparse adagrad in https://github.com/pytorch/pytorch/blob/master/torch/optim/adagrad.py#L73-L77. I will change both places.

[the remark this answers is not in the document]
Comment (SsnL) [reply torch/optim/lazy_adam.py]: There is an issue with `cuda_sparse_tensor.new`. I'm investigating it. Afterwards I'll update this PR.

**Comment by apaszke** [comment]
Actually how about calling it `SparseAdam`? It's probably going to make it a lot easier for people to find. It's not really lazy in any sense, it just applies updates sparsely.

**Review by ezyang — CHANGES_REQUESTED** [review]
C.f. Adam's comments

**Comment by SsnL** [comment]
Yeah this is still WIP in current state. I'll get back to this after fixing the coalesce bug & finish the sparse dataparallel thing.

File: torch/optim/lazy_adam.py (hunk 3 above)
Comment (SsnL) [reply torch/optim/lazy_adam.py]: Yes! It is extracting `dense._sparse_mask(sparse)` extracts dense values at `sparse._indices()`. So it needs to be called multiple times.

**Comment by SsnL** [comment]
cc @gchanan this is the "sparse version of adam"

**Comment by SsnL** [comment]
Request a review :) Thanks in advance!
@ezyang @apaszke

File: torch/optim/sparse_adam.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+                # Decay the first and second moment running average coefficient
+                #      old <- b * old + (1 - b) * new
+                # <==> old += (1 - b) * (new - old)
+                old_exp_avg_values = exp_avg._sparse_mask(grad)._values()
+                exp_avg.add_(make_sparse(grad_values.sub(old_exp_avg_values).mul_(1 - beta1)))
+                old_exp_avg_sq_values = exp_avg_sq._sparse_mask(grad)._values()
+                exp_avg_sq.add_(make_sparse(grad_values.pow(2).sub_(old_exp_avg_sq_values).mul_(1 - beta2)))
+
+                numer = exp_avg._sparse_mask(grad)._values()
```
Comment (apaszke) [inline torch/optim/sparse_adam.py]: I guess you could still avid these two `_sparse_mask`s if only you did the update more like this (dense add is likely to be faster than sparse mask, although I didn't check):
```python
old_exp_avg_values = exp_avg._sparse_mask(grad)._values()
exp_avg_update = grad_values.sub(old_exp_avg_values).mul_(1 - beta1)
exp_avg.add_(make_sparse(exp_avg_update))
# similarly for exp_avg_sq

numer = exp_avg_update.add_(old_exp_avg_values)
# similarly for denom
del exp_avg_update, exp_avg_sq_update # it was modified, so be safe and remove the reference
```

**Review by apaszke — APPROVED** [review]
LGTM. It might be possible to improve efficiency a bit by reducing the number of masking operations, but it needs some further benchmarks. I haven't verified the math carefully

**Comment by SsnL** [comment]
Thanks Adam. I updated the PR.

File: torch/optim/sparse_adam.py [hunk 7]
```diff
@@ -75,17 +75,22 @@ def make_sparse(values):
...
+                exp_avg_update_values = grad_values.sub(old_exp_avg_values).mul_(1 - beta1)
+                exp_avg.add_(make_sparse(exp_avg_update_values))
                 old_exp_avg_sq_values = exp_avg_sq._sparse_mask(grad)._values()
-                exp_avg_sq.add_(make_sparse(grad_values.pow(2).sub_(old_exp_avg_sq_values).mul_(1 - beta2)))
+                exp_avg_sq_update_values = grad_values.pow(2).sub_(old_exp_avg_sq_values).mul_(1 - beta2)
+                exp_avg_sq.add_(make_sparse(exp_avg_sq_update_values))
 
-                numer = exp_avg._sparse_mask(grad)._values()
-                denom = exp_avg_sq._sparse_mask(grad)._values().sqrt_().add_(group['eps'])
+                # Dense addition again is indended, avoiding another _sparse_mask
```
Comment (apaszke) [inline torch/optim/sparse_adam.py]: nit: indended

File: torch/optim/sparse_adam.py [hunk 8]
```diff
@@ -75,17 +75,22 @@ def make_sparse(values):
...
+                numer = exp_avg_update_values.add_(old_exp_avg_values)
+                denom = exp_avg_sq_update_values.add_(old_exp_avg_sq_values).sqrt_().add_(group['eps'])
 
                 bias_correction1 = 1 - beta1 ** state['step']
                 bias_correction2 = 1 - beta2 ** state['step']
                 step_size = group['lr'] * math.sqrt(bias_correction2) / bias_correction1
 
                 p.data.add_(make_sparse(-step_size * numer.div_(denom)))
 
+                del exp_avg_update_values, exp_avg_sq_update_values
```
Comment (apaszke) [inline torch/optim/sparse_adam.py]: oh the del was only meant to be close to the place where we overwrite the `update_values`, because they are no longer what the name suggests. It's not needed if you put it at the end

Comment (SsnL) [reply torch/optim/sparse_adam.py]: Oh gotcha! :)

File: torch/optim/sparse_adam.py (hunk 7 above)
Comment (SsnL) [reply torch/optim/sparse_adam.py]: ;) thanks

**Comment by octavian-ganea** [comment]
Is there any plan to make SGD with weight decay and ASGD work with sparse gradients ? 

Thanks,

**Comment by souryadey** [comment]
I am wondering why there is no weight_decay argument for SparseAdam? I understand that the parameter tensors are dense and the grads are sparse, so this line from regular Adam will not work:
`grad.add_(group['weight_decay'], p.data)`

How about only applying weight decay to those elements in `p` which correspond to the nnz in `grad`? Something like:
`grad.add_(group['weight_decay'], p.sparse_mask(grad).data)`

**Comment by soumith** [comment]
@souryadey because then the weight decay would be extremely weird. Elements that are frequently accessed will have significant weight decay applied to them, whereas elements which are barely accessed will have almost no weight decay applied. That would slow down learning and is not very well defined mathematically either, i think.

**Comment by souryadey** [comment]
@soumith It seems to me that what you are suggesting is the opposite of pruning. Elements which are important and should have large weight values will have their weights decayed, while elements which are not accessed often (whose weight values should be small) will not have any decay, so essentially there will be no difference between important and unimportant weights. Hope this understanding is correct.

I'm wondering in what kind of situations will Sparse Adam be useful (i.e. grads will be sparse)?
